spring-session/src/main/java/org/springframework/session/IndexResolver.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2014-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.session;
+
+/**
+ * Strategy interface for resolving the value of the {@link Session}'s index.
+ *
+ * @param <S> the type of Session being handled
+ * @author Vedran Pavic
+ * @since 1.3.0
+ * @see FindByIndexNameSessionRepository
+ */
+public interface IndexResolver<S extends Session> {
+
+	/**
+	 * Resolve the value of the given index from the session.
+	 * @param session the session
+	 * @param indexName the index name to resolve
+	 * @return the resolved index value, {@code null} if none found
+	 */
+	String resolve(S session, String indexName);
```

Comment:
> either using a CompositeIndexResolver or some other way.

This requires another API because the current interface must have knowledge of the index name.

> If you would go the input should just be the session and the result that has a mapping of index name to index value route then you wouldn't be able to reuse this logic.

Why not? Seems you could do this:

```java

public interface IndexResolver<S extends Session> {
	Map<String,String> resolveIndexes(S session);
}

public class PrincipalNameIndexResolver<S extends Session> implements IndexResolver<S> {
	private SpelExpressionParser parser = new SpelExpressionParser();
	private String indexName = FindByIndexNameSessionRepository.PRINCIPAL_NAME_INDEX_NAME;

	public String resolvePrincipal(Session session) {
		String principalName = session.getAttribute(FindByIndexNameSessionRepository.PRINCIPAL_NAME_INDEX_NAME);
		if(principalName != null) {
			return principalName;
		}
		Object authentication = session.getAttribute("SPRING_SECURITY_CONTEXT");
		if(authentication != null) {
			Expression expression = parser.parseExpression("authentication?.name");
			return expression.getValue(authentication, String.class);
		}
		return null;
	}

	@Override
	public Map<String, String> resolveIndexes(S session) {
		String index = resolvePrincipal(session);
		if(index == null) {
			return Collections.emptyMap();
		}
		return Collections.singletonMap(indexName, index);
	}
}

public class GemfireIndexResolver<S extends Session> implements IndexResolver<S> {
	private PrincipalNameIndexResolver<S> principalResolver = new PrincipalNameIndexResolver<>();

	@Override
	public Map<String, String> resolveIndexes(S session) {
		Map<String, String> gemfireIndexes = resolveThem(session);
		gemfireIndexes.putAll(principalResolver.resolveIndexes(session));
		return gemfireIndexes;
	}

	private Map<String, String> resolveThem(S session) {
		// TODO
		return null;
	}
}
```